File: lc/py/019.py
# 1D DP
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [1] + [0] * (n-1)

        for _ in range(m):
            for c in range(1, n):
                dp[c] = dp[c] + dp[c-1]
        return dp[-1]

# 2D DP
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0] * n for _ in range(m)]
        dp[0][0] = 1

        for r in range(m):
            for c in range(n):
                top = dp[r-1][c] if r > 0 else 0
                left = dp[r][c-1] if c > 0 else 0
                dp[r][c] = dp[r][c] + top + left
        return dp[-1][-1]

File: lc/c/019.c
// 1D DP
int uniquePaths(int m, int n){
    int dp[n];
    memset(dp, 0, sizeof dp);
    dp[0] = 1;

    for (int r = 0; r < m; ++r) {
        for (int c = 1; c < n; ++c)
            dp[c] += dp[c-1];
    }
    return dp[n-1];
}

// 2D DP
int uniquePaths(int m, int n){
    int dp[m][n];
    memset(dp, 0, sizeof dp);
    dp[0][0] = 1;

    for (int r = 0; r < m; ++r) {
        for (int c = 0; c < n; ++c) {
            int top = r > 0 ? dp[r-1][c] : 0;
            int left = c > 0? dp[r][c-1] : 0;
            dp[r][c] = dp[r][c] + top + left;
        }
    }
    return dp[m-1][n-1];
}
